🐛 BUG BASH 🐛 Add: initial new tutorial - what is a Python package + images ✨ (#117)
pyOpenSci/python-package-guide: tutorials/intro.md

> # Python packaging 101

Author: ALL: please review this file. the index file changes above just add relevant navigation to get to this tutorial.

            </div>
        </div>
    </div>
</div>
-->

<!--

TODO: whatever graphic we use we can then add it to the top of each page to show the user where they are in the progression
* Work on the create a license lesson next?? that is the most incomplete
* then start the conda one and place the pyos package on conda-forge??
* Below i have an image overview of the lessons and a list of links. i'm thinking a visual html / css block that is clickable would be a more powerful "graphic". it could still have colors.
-->

Welcome to the pyOpenSci Python packaging tutorial series. The lessons on the upcoming pages walk you through the core steps needed to
create a Python package.

:::{figure-md} packaging-outline

<img src="../images/tutorials/packaging-101-outline.png" alt="Diagram showing .. more here if this stays." width="800px">

TODO: This will be replaced with a interactive CSS element that walks users through lessons
:::

## Who are these tutorials for?

The content in this tutorial series is beginner friendly and assumes
that you have not created a Python package before. However, the
content will still be valuable if you are interested in better

> understand the steps for creating a Python package.

Reviewer: ```suggestion
understanding the steps involved in creating a Python package.
```

<!-- it might be interesting to make the above graphic a css
set of blocks that are colored. then you could hover over each on to get to the lesson of choice. this would make it easier to also
customize each page with that block at the top -->

> In this series you will learn about the core elements that you need to publish your package to PyPI.

Reviewer: ```suggestion
In this series you will learn about the core elements that you need to publish your package to the [Python Package Index (PyPI)](https://pypi.org/).
```

```{toctree}
:hidden:
:caption: Python Packaging 101

What is a Python package? <self>
Make your code pip installable <1-installable-code>
Add README file <2-add-readme>
Add LICENSE & CODE of CONDUCT <4-add-license-file>
Add pyproject.toml <5-pyproject-toml>
Publish to PyPI <6-publish-pypi>
Publish to conda-forge <7-publish-conda-forge>

```

## What is a Python package?

:::{admonition} Learning Objectives

This lesson introduces you to the basic components of a Python package.
After reading this lesson you will:

- Understand what a Python package is
- Be able to list the 5 core components of a Python package
- You will be able to explain the difference between generalizable code and code that supports a specific scientific application

:::

A Python package is a collection of related modules / code containing
functions, classes and methods that are organized together in a directory.
Packages allow you to group and structure your Python code, making it easier
to manage and reuse code across different projects.

A package is installable which means that you can add the functionality within
the package code to any Python environment and import that functionality like

> you would import numpy or matplotlib.

Reviewer 2: I was mainly looking for consistency, agree we should match existing documentation. 

And yes on a style guide!

> you would import numpy or matplotlib.

Author: i also see they use NumPy not numpy. so avoid code formatting for package names - is that the consensus here? i'm all for a style guide as well.

> you would import numpy or matplotlib.

Author: i've removed code formatting for all package names (i think) locally now!

At a high level, you can think about each package as a toolbox filled with
different tools that perform specific actions in your code when imported and

> called.

Author: excellent! i like the restructuring and threading the tool box analogy throughout.

:::{figure-md} python-toolbox

<img src="../images/tutorials/toolbox.png" alt="Diagram showing ADD ALT." width="300px">

You can think about a package as a toolbox filled with coding tools. A tool may be a function or a
class. Each tool does a specific thing well.
:::

<!--
:::{admonition} TODOS
:class: todo
TODOs:
* Create a nice graphic for this landing page that introduces the pieces of a python package and also how these tutorials are organized ... maybe even that step box diagram?

- sometimes people might start by creating a new repo and cloning it.
* It would be good to have them add development requirements up front in their pyproject.toml file
* Other times they might start by working locally and then pushing the repo content to github / gitlab. in any case the readme and license lessons may or may not be relevant. so we should have a pregame for setting up a fresh repo, creating an initial .gitignore, readme and license. and then link to the chose a license lesson.

::: -->

## The elements of a Python package

:::{figure-md} package-components

<img src="../images/tutorials/package-components.png" alt="Diagram showing .. more here if this stays." width="500px">

> The elements of a Python package include code, documentation, tests a OSI-approved license and a specific directory structure. Maintainers are at the core making sure everything works and is up to date while fixing bugs and addressing user concerns.

Author: fixed!! Jeremiah i'm going to add thumbs up emojis to each of these changes as i fix it locally!! then i'll resolve :) but i wanted to comment on this one to thank you for all of the feedback before reverting to emoji style feedback!

:::

The core elements of Python package include:

- **Code:** Generalized code that performs operations that you may need to complete multiple times
- **Documentation:** documentation with tutorials / examples that help users get started using your tool
  - Documentation also helps people to contribute to your package.
- **Tests:** that makes sure your code works as it should and makes it easier for you and others to contribute to, modify and update the code in the future
- **License:** An open source license …. - link to choose a license…

> - **Infrastructure** that automates updates, publication workflows and runs test suites

Author: great suggestion! i see the inconsistency. 

let's explore this - are there 5 elements and what should they be?

1. docs - contributor / development + user
2. code
3. tests (also code but keeping it separate as it's often something scientists add after even tho test based design is a thing) 
4. License
5. Infrastructure (rather than file structure) - we could argue that all of the above things need to be organized in a specific way for a user to build & publish their package successfully.

> If you intend for others to use and contribute to your code, consider who will maintain it over time. You will want a **contributing / development** guide to help new potential contributors get started with contributing to your package. And a **code of conduct** to ensure community interactions remain healthy both for you and your contributors / maintainer team

Reviewer 2: ```suggestion
If you intend for others to use and contribute to your code, consider who will maintain it over time. You will want a **contributing / development** guide to help new potential contributors get started with contributing to your package, as well as a **code of conduct** to ensure community interactions remain healthy both for you and your contributors / maintainer team
```

:::{admonition} What pyOpenSci looks for in a package
:class: tip

pyOpenSci performs an [initial set of editor checks](https://www.pyopensci.org/software-peer-review/how-to/editor-in-chief-guide.html#editor-checklist-template) for any package submitted
to us for peer review.
You may find these checks useful as you create your package as a baseline for
things that you package should have.

:::

## Packages are more than just code

A package in any language is more than just code. If your package is public facing, meaning people besides yourself will use it, you should consider the various elements of a package that make it a useful community resource.

### Version control and storing your package on GitHub or GitLab

Most Python packages live in an online version control platform such as GitHub
or GitLab. GitHub and GitLab are both online platforms that
run [git](https://git-scm.com/) for version control. Have your software
under version control is important because it allows you to both track changes over time while also going back in history and undoing changes in the case that a change to the code base unexpectedly breaks something.

> By publishing your package on GitHub or GitLab, you are now making your code public facing. this means that others can both see your code and also make contributions using a pull request - review workflow.

Author: and i learned something about GitLab - thank you!

:::{admonition} GitHub & GitLab vs. Git
:class: tip

> GitHub and GitLab are online (cloud) platforms that run git (version control

Reviewer 2: ```suggestion
GitHub and GitLab are online (cloud) platforms that run `git` (version control
```
Consistent monospace

software) on the backend. Running git locally on your computer allows you to
upload (`git push`) and download (`git pull`) files to GitHub and GitLab.
:::

### Continuous integration and continuous deployment help you maintain your package

> Platforms such as GitHub and GitLab also provide continuous integration and continuous deployment (CI/CD). Continuous integration (CI) refers to a platform that automatically runs a specific job when a certain event occurs.

Reviewer 3: I think it would also be good to mention ticket trackers. This is another service provided by github and gitlab but is really totally separate from git. It also ties in strongly with users and contributors as a contributor guide might mention how to open a ticket or a PR and what is expected of both the author and the maintainer when either is created. I'm sure going over everything could even be its own section (probably part of "Infrastructure").

Author: oK!! i'm still questioning the diagram BUT i've added a "infrastructrure" section locally with an H2 tag. within that is github/gitlab, ticket trackers and ci/cd 

none of that is discussed in detail in this specific lesson series but we can add language around that.

**An example of Continuous integration:**
* When someone submits a change to your code, your tests will run across different operating systems and the code will be checked for format issues.

**An example of Continuous deployment:**
* When you are ready to release your package to PyPI, a continuous deployment operation might be triggered on release to publish your package to PyPI.

> Integrated CI/CD will help you maintain  your software ensuing that changes to the code don't break things unexpectedly and also maintain a style and format consistency.

Author: Fixed!

:::{figure-md} packaging-workflow

<img src="../images/tutorials/packaging-lifecycle.png" alt="Diagram showing .. more here if this stays." width="700px">

The lifecycle of a scientific Python package.
:::

<!-- >then in the publish section below we can add the arrow with that structure being
>built into distributions that get published on PyPI and conda...

>then in the community part we add another section that is the github repo with users contributing... very simple diagrams... -->

```{toctree}
:hidden:
:caption: Pre game
:glob:
extras/*

```

## What should code in a Python package look like?

> Ideally the code in your Python package is general. This means it

Reviewer 4: I'm afraid that as written this could give a newbie the impression that they should only make a package if they've developed a new matplotlib or something similarly broad.

I know it's written to try and avoid that, but I think it's just really hard to define "general" and "generalizable".

Here's three suggesions:
1. rephrase to talk about a Python package as "having the potential to be used by more than one project" or something a little less daunting
2. acknowledge that it's not easy to figure out when a package can be re-used, but immediately contrast that with code written *specifically* for a research project, such as a script that carries out only one analysis in that project, and is meant to only operate on a very specific set of pre-processed data --maybe the "how does this relate to research compendia" box could move up here? Maybe have the text in-line to contrast "general package" with "code for a research project" and then just have the call-out say "by the way, just like there are good practices for structuring your Python package, there are good practices for structuring code and data for a research project as discussed in this paper (link)--in fact some researchers specifically study how to structure what they call research compendia (link) or repro-packs (link)"
3. Sprinkle current pyOpenSci packages throughout as examples of more general packages

Reviewer 4: One last comment here:
you could also say something like "often you will find after you have written code for several projects that you have an idea for a new, more general Python package that would have made those projects easier!"

Author: this feedback is so helpful @NickleDave i'll work on finessing this section some to make it less daunting. i can see your perspective of how someone might read it that way!!

Author: ok i believe i've addressed most of the above locally!! will push it up shortly. moving up that research compendia section is a great call - it fits well. i also created a subsection around package scope. that avoids the more generic / generalized language which might confuse users!

can be used on different data or for different scientific applications. An example
of a package that is written in a generalized way is `matplotlib`.

`matplotlib` does
one (big important) thing really well:

*It creates visual plots of data.*

> `Matplotlib` is used by thousands of users for different plotting applications

Reviewer 2: `matplotlib` starts both these sentences, but is capitalized in only one of them. Should be consistent.

idea of code being used for something more than a single workflow still applies
to package development if you want other people to use your package.

### Code should also be clean & readable & documented

The code in your package should also be clean, readable and well documented.

**Clean code:** Clean code refers to code that uses expressive variable names,
is concise and does not repeat itself. We will dive deeper into best practices
for clean code in future pyOpenSci tutorials.

**Readable code:** Readable code is code written with a consistent style.
You can use linters and code formatters such as black and flake8 to ensure
this consistency throughout your entire package. [Learn more about code formatters here.](../package-structure-code/code-style-linting-format.html)

**Documented code:** documented code is written using docstrings that help a
user understand both what the functions and methods in your code does and also
what the input and output elements of each function is. [You can learn more about docstrings in our guide, here.](../documentation/write-user-documentation/document-your-code-api-docstrings)

:::{admonition} Where do research compendia fit in?
:class: note

> A Research Compendium is an organized set of code, data and documentation that

Reviewer 2: I haven't heard the specific term "research compendium" (rather than just "research code"). Is there a source for this use?

Reviewer 2: I agree, this extra context would be great!

Author: i forgot about that paper!! starring all of our favorite people (several of whome are in our leadership at pyOpenSci!! yay. ok i've made a bunch of changes here now. 

1. i moved this up to a section on package scope and various reasons to publish a package
2. reasons include both a package that you just need for your own projects - that might be pip installabale. it's still a package but doesn't need contributing docs and tutorials necessarily. and packages you expect others to use that then need additional documentation and infra
3. i've renamed the admonition to make it easier to understand and added several links around the idea of research compendia!! 

i'm grateful for this discussion - thank you both!

supports a specific research project. It aims to enhance the reproducibility and
transparency of research by providing a comprehensive record of the methods,
data, and analyses used in a study.

A Python package however is a collection of modules that are focused on a
specific set of tasks that can be applied across numerous research projects.

> As such a Python package is more generalizable than a research Compendium

Reviewer 2: ```suggestion
As such a Python package is more generalizable than a Research Compendium
```
should be consistent with capitalization

:::

## Making your package installable - publishing to PyPI & conda-forge

### Python packages and environments

A Python package can be installed into a Python environment in the same way
you might install `numpy` or `pandas`. Installing your package into an environment
allows you to access it from any code run with that specific Python environment activated.

:::{figure-md} packages-environment

<img src="../images/tutorials/environment-package-install.png" alt="Diagram showing the steps associated with creating a package and then installing it. The first arrow says your package and the second says pip install package. The second arrow leads to a box that represents a python environment that already has some packages installed such as pandas and numpy. Your package will also get installed into that same environment when you pip install it." width="700px">

You don't have to publish to PyPI in order to make your code installable.
WIth the correct file structure and project metadata you can make your code
installable locally on your computer and use it for projects that you are working
on without having to ever publish to PyPI. Publishing to PyPI is useful when you

> want to make your code public facing and share it with others.

Reviewer 2: ```suggestion
want to make your code public-facing and share it with others.
```

:::

### Publishing a package

> If you want to make your package installable using `pip` you need to

Author: i can see how it's confusing. i've tried to reword so it's a bit more clear that it's installable without needing to download the code (having the source locally).

publish it in a repository such as PyPI or conda-forge. If you package is a
pure python package, then publishing to both PyPI and conda-forge only takes just a few steps.

In these tutorials, you will learn [how to publish to PyPI using twine.](6-publish-pypi.md)

Then you can create a conda-forge recipe using the grayskull tool. This recipe can then be submitted to conda-forge. [You will learn more about the conda-forge publication process here.](7-publish-conda-forge.md)

> :::{figure-md} build-workflow-tutorial

Author: oh you are right. let's fix this.  

would [linking to this section of our guide help](https://www.pyopensci.org/python-package-guide/package-structure-code/python-package-distribution-files-sdist-wheel.html#source-distribution-sdist) in the caption of the image? i totally get that. i wouldn't know what those terms were either if i was a beginner (and i didn't for a long time) @billbrod ?? 

also wold something like (package distribution files (sdist & wheel) help in the image itself? 

i'll try both but if other fixes would be helpful please share!!

> :::{figure-md} build-workflow-tutorial

Author: ok i've made a few changes and you can have a look

1. i'd added shortcuts to both the sdist and wheel discussion pages for easy linking!!
2. i've added links to both of those distribution file descriptions in the caption of the image and described what they are. 

hopefully this clarifies things a lot but i'll be curious to see know what you think !

<img src="../images/tutorials/publish-package-pypi-conda.png" alt="Graphic showing the high level packaging workflow. On the left you see a graphic with code, metadata and tests in it. those items all go into your package. Documentation and data are below that box because they aren't normally published in your packaging wheel distribution. an arrow to the right takes you to a build distribution files box. that box leads you to either publishing to testPyPI or the real pypi. from PyPI you can then connect to conda forge for an automated build that sends distributions from PyPI to conda-forge." width="700px">

> Once you are ready to make your code public, you can publish it on PyPI. Once the code is on PyPI it's a simple set of steps to publish to conda forge. You create a recipe using the grayskull package and then you open a pr in the conda-forge recipe repo. You will learn more about this process in the [conda-forge lesson](#).

Author: i will come back to this as i think we need consensus around package names being within code brackets (monospace/ highlighted) vs not. i will open an issue about style guide items that we can track. as once we decide it's not hard to do find and replaces across the entire guide. and i think there are even tools to help with style guide enforcement that we could add to ci.

> Once you are ready to make your code public, you can publish it on PyPI. Once the code is on PyPI it's a simple set of steps to publish to conda forge. You create a recipe using the grayskull package and then you open a pr in the conda-forge recipe repo. You will learn more about this process in the [conda-forge lesson](#).

Reviewer 2: As an aside: I'd be very interested in learning more about style guide enforcement tools! Would a discussion of useful CI tools for python packaging tools belong in pyOpenSci? Asking for myself, because I'd love to have a list of them.

:::

## Yay, your package has users! Now what?

As the community using your package grows, you may also find yourself managing users, contributors and others who want to interact with your package. It’s important to consider all this before you dive into development. Once you have a user base in the community, people will depend upon your code to work and will need direction regarding how to use it.

To support your community, you'll want to add things like:

* development guide
* code of conduct
* contributing guide

> 

Author: better yet - we have pages in our guide for these things! i'll link to each and if we need to update those pages as well in a future PR we can. have a look if you wish here for instance is the contributing file!! great suggestion. the idea here just so you know Bill is that we want to make use of the overview content in our guide - in the tutorials (and vice versa). so all of this is intertwined. but these pages are how to pages. the guidebook pages provide overviews of concepts, tools, etc!!

https://www.pyopensci.org/python-package-guide/documentation/repository-files/contributing-file.html 

i'll make these fixes now!

> 

Author: excellent suggestion btw - i just provided the above context so YOU knew we had other content as this is only a single page. but we DO want to cross link so that context that you are pointing out was missed by me AND is super important
